tree/ntuple/v7/inc/ROOT/RMiniFile.hxx

```diff
@@ -0,0 +1,221 @@
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RMiniFile
+#define ROOT7_RMiniFile
+
+#include <ROOT/RNTupleOptions.hxx>
+#include <ROOT/RStringView.hxx>
+
+#include <cstdint>
+#include <cstdio>
+#include <memory>
+#include <string>
+
+class TFile;
+
+namespace ROOT {
+namespace Experimental {
+
+namespace Detail {
+class RRawFile;
+}
+
+// clang-format off
+/**
+\class ROOT::Experimental::RNTuple
+\ingroup NTuple
+\brief Entry point for an RNTuple in a ROOT file
+
+The class points to the header and footer keys, which in turn have the references to the pages.
+Only the RNTuple key will be listed in the list of keys. Like TBaskets, the pages are "invisible" keys.
+Byte offset references in the RNTuple header and footer reference directly the data part of page records,
+skipping the TFile key part.
+*/
+// clang-format on
+struct RNTuple {
+   /// Allows for evolving the struct in future versions
+   std::uint32_t fVersion = 0;
+   /// Allows for skipping the struct
+   std::uint32_t fSize = sizeof(RNTuple);
+   /// The file offset of the header excluding the TKey part
+   std::uint64_t fSeekHeader = 0;
+   /// The size of the compressed ntuple header
+   std::uint32_t fNBytesHeader = 0;
+   /// The size of the uncompressed ntuple header
+   std::uint32_t fLenHeader = 0;
+   /// The file offset of the footer excluding the TKey part
+   std::uint64_t fSeekFooter = 0;
+   /// The size of the compressed ntuple footer
+   std::uint32_t fNBytesFooter = 0;
+   /// The size of the uncompressed ntuple footer
+   std::uint32_t fLenFooter = 0;
+   /// Currently unused, reserved for later use
+   std::uint64_t fReserved = 0;
+
+   bool operator ==(const RNTuple &other) const {
+      return fVersion == other.fVersion &&
+         fSize == other.fSize &&
+         fSeekHeader == other.fSeekHeader &&
+         fNBytesHeader == other.fNBytesHeader &&
+         fLenHeader == other.fLenHeader &&
+         fSeekFooter == other.fSeekFooter &&
+         fNBytesFooter == other.fNBytesFooter &&
+         fLenFooter == other.fLenFooter &&
+         fReserved == other.fReserved;
+   }
+};
+
+namespace Internal {
+
+/// Holds status information of an open ROOT file during writing
+struct RTFileControlBlock;
+
+// clang-format off
+/**
+\class ROOT::Experimental::Internal::RMiniFileReader
+\ingroup NTuple
+\brief Read RNTuple data blocks from a TFile container, provided by a RRawFile
+
+A RRawFile is used for the byte access.  The class implements a minimal subset of TFile, enough to extract
+RNTuple data keys.
+*/
+// clang-format on
+class RMiniFileReader {
```

Comment:
Is it worth noting in the name that this reads only RNTuples and/or will that class be extended later to support more types?

Reply:
I'm afraid it becomes a bit longish... Perhaps more generally we should discuss an RNTuple namespace (like `ROOT::RDF`) because there are more classes with generic names (e.g. `RPage`). How about `ROOT::RNTPL`?

Reply:
(Regarding the other question: there are no immediate plans to extend it but I wouldn't exclude that option either.)

Reply:
>  we should discuss an RNTuple namespace 

What would be in there?

Reply:
I have in mind to put in the generic classes that make only sense in RNTuple context: `RPage`, `RField`, `RValue`, `RPageStorage`, `RView`.  Some of the classes are currently prefixed by "NTuple", such as `RNTupleView`. Perhaps we can benefit for RDF experience for the namespace shuffle.